url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/KitRobotAgent.java

```diff
@@ -0,0 +1,461 @@
+import agent.interfaces.Stand;
+import agent.test.mock.MockGraphics;
+
+/**
+ * Kit Robot brings moves kits to and from the conveyor and arranges kits on the kitting stand. It is responsible for
+ * moving the assembled kits on the stand into the inspection area for the Camera. Interacts with the stand, Conveyor
+ * and Camera.
+ * 
+ * @author Daniel Paje
+ */
+public class KitRobotAgent extends Agent implements KitRobot {
+
+	private final List<MyKit> myKits = Collections.synchronizedList(new ArrayList<MyKit>());
+
+	// Tracks stand positions and whether or not they are open
+	Map<Integer, Boolean> standPositions = Collections.synchronizedMap(new TreeMap<Integer, Boolean>());
+
+	private boolean kitWaitingOnConveyor;
+	private boolean kitRequested;
+	private int numKitsToMake;
+	private int numKitsRequested;
+
+	// Used to prevent animations from overlapping
+	Semaphore animation = new Semaphore(0, true);
+
+	// References to other agents
+	private Stand stand;
+	private Conveyor conveyor;
+	private Camera camera;
+	private KitRobotGraphics kitrobotGraphics;
+	private MockGraphics mockgraphics;
+
+	private KitRobotState state;
+
+	private enum KitRobotState {
+		IDLE, HOLDING_KIT, NOT_HOLDING_KIT
+	};
+
+	private final String name;
+
+	/**
+	 * Inner class encapsulates kit and adds states relevant to the stand
+	 * 
+	 * @author dpaje
+	 */
+	public class MyKit {
+		public Kit kit;
+		public KitStatus KS;
+		public int location;
+
+		public MyKit(Kit k) {
+			this.kit = k;
+			this.KS = KitStatus.AWAITING_PICKUP;
+		}
+	}
+
+	public enum KitStatus {
+		AWAITING_PICKUP, REQUESTED, PICKED_UP, ON_STAND, MARKED_FOR_INSPECTION, AWAITING_INSPECTION, FAILED_INSPECTION, PASSED_INSPECTION, SHIPPED;
+	};
+
+	/**
+	 * Constructor for KitRobotAgent class
+	 * 
+	 * @param name
+	 *            name of the kitrobot
+	 */
+	public KitRobotAgent(String name) {
+		super();
+
+		this.name = name;
+		kitWaitingOnConveyor = false;
+		kitRequested = false;
+		numKitsRequested = 0;
+		numKitsToMake = 0;
+		state = KitRobotState.IDLE;
+
+		// Don't assume stand is empty
+		standPositions.put(0, true);
+		standPositions.put(1, false);
+		standPositions.put(2, false);
+	}
+
+	/*
+	 * Messages
+	 */
+
+	@Override
+	public void msgNeedThisManyKits(int total) {
+		print("Received msgNeedThisManyKits with amount " + total);
+		conveyor.msgNeedThisManyKits(total);
+		numKitsToMake = total;
+		numKitsRequested = 0;
+		kitRequested = false;
+		stateChanged();
+	}
+
+	@Override
+	public void msgKitReadyForPickup() {
+		print("Received msgKitReadyForPickup");
+		kitWaitingOnConveyor = true;
+		stateChanged();
+	}
+
+	@Override
+	public void msgNoKitsLeftOnConveyor() {
+		print("Received msgNoKitsLeftOnConveyor");
+		kitWaitingOnConveyor = false;
+		// stateChanged();
+	}
+
+	@Override
+	public void msgHereIsKit(Kit k) {
+		print("Received msgHereIsKit");
+		print("Adding kit " + k.toString());
+		kitRequested = false;
+		MyKit mk = new MyKit(k);
+		myKits.add(mk);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **132**. At this line the following field is not protected by synchronization: `myKits`. Your code typically protects this field with synchronization on object: `myKits`.

**Fix**  
 Consider protecting the above field with synchronization on object: `myKits`. For example, like at lines: **166**, **180**, and **255**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **149**, **150**, **166**, **180**, **222**, **235**, **255**, **269**, and **376**.